tree/ntupleutil/v7/src/RNTupleImporter.cxx

```diff
@@ -13,16 +13,331 @@
+      if (isLeafList && isCountLeaf)
+         return R__FAIL("unsupported: count leaf arrays in leaf list, branch " + std::string(b->GetName()));
+
+      // Only plain leafs with type identifies 'C' are C strings. Otherwise, they are char arrays.
+      Int_t firstLeafCountval;
+      const bool isCString = !isLeafList && (firstLeaf->IsA() == TLeafC::Class()) &&
+                             (!firstLeaf->GetLeafCounter(firstLeafCountval)) && (firstLeafCountval == 1);
+
+      if (isCountLeaf) {
+         // This is a count leaf.  We expect that this is not part of a leaf list. We also expect that the
+         // leaf count comes before any array leaves that use it.
+         // Count leaf branches do not end up as (physical) fields but they trigger the creation of an untyped
+         // collection, together the collection mode.
+         RImportLeafCountCollection c;
+         c.fCollectionModel = RNTupleModel::CreateBare();
+         c.fMaxLength = firstLeaf->GetMaximum();
+         c.fCountVal = std::make_unique<Int_t>(); // count leafs are integers
+         // Casting to void * makes it work for both Int_t and UInt_t
+         fSourceTree->SetBranchAddress(b->GetName(), static_cast<void *>(c.fCountVal.get()));
+         // We use the leaf pointer as a map key.  The array leafs return that leaf pointer from GetLeafCount(),
+         // so that they will be able to find the information to attach their fields to the collection model.
+         fLeafCountCollections.emplace(firstLeaf->GetName(), std::move(c));
+         continue;
+      }
+
+      std::size_t branchBufferSize = 0; // Size of the memory location into which TTree reads the events' branch data
+      // For leaf lists, every leaf translates into a sub field of an untyped RNTuple record
+      std::vector<std::unique_ptr<Detail::RFieldBase>> recordItems;
+      for (auto l : TRangeDynCast<TLeaf>(b->GetListOfLeaves())) {
+         if (l->IsA() == TLeafObject::Class()) {
+            return R__FAIL("unsupported: TObject branches, branch: " + std::string(b->GetName()));
+         }
+
+         Int_t countval = 0;
+         auto *countleaf = l->GetLeafCounter(countval);
+         const bool isLeafCountArray = (countleaf != nullptr);
+         const bool isFixedSizeArray = (countleaf == nullptr) && (countval > 1);
+
+         std::string fieldName = isLeafList ? l->GetName() : b->GetName();
+         std::string fieldType = isCString ? "std::string" : l->GetTypeName();
+         if (isFixedSizeArray)
+            fieldType = "std::array<" + fieldType + "," + std::to_string(countval) + ">";
+
+         RImportField f;
+         f.fIsClass = isClass;
+         std::unique_ptr<Detail::RFieldBase> field;
+         if (isCString) {
+            branchBufferSize = l->GetMaximum();
+            auto result = Detail::RFieldBase::Create(fieldName, fieldType);
+            if (!result)
+               return R__FORWARD_ERROR(result);
+            field = result.Unwrap();
```

Comment:
Isn't this part common also to the `else` part of the `if` statement?

Reply:
Yep, factored out.